### ISSUE TITLE
[WasmFS+OPFS] Unspecified file size limit

### ISSUE THREAD

**Issue by DenizUgur** [emscripten-core/emscripten#21335]
I have a wasm file compiled with `-lopfs.js -sWASMFS -sFORCE_FILESYSTEM` and I'm trying to perform IO operations on a large (4GB) file. The process works properly for smaller files (~150MB, probably anything lower than `2^32-1` bytes). However, I've noticed that when I try to read this large file from WASM it fails. When I try to read it using `readFile` from JS I don't see correct size on ArrayBuffer. I don't expect to see all of them (ArrayBuffer limit) but it's way less than my browser's limit (Chrome).

In essence, I'm trying to pass a `File` provided by user to `OPFS` and use it with my executable. `WORKERFS` works very nice if I create the node with `WORKERFS.createNode`. However, the problem with that is reading back the data if it's larger than `2^32-1` bytes. I'm considering overloading `WORKERFS.stream_ops.write`/`FS.write` to write to `OPFS` and create `File` object using that.

> I've verified that the file that has been written to OPFS was correct by downloading it again.

A wild guess but maybe **seeking more than `2^32-1` bytes** at once is creating this problem?

EDIT: I've confirmed this is the problem. I had a MP4 file intentionally made larger than it's supposed to be. This is done by marking first sample at an offset of ~4GB. Then I've tried a normal 4GB downloaded from YouTube and it worked properly. So it's not size issue but rather seeking issue.

What's your opinion on this process? Is there really a file size limit with `OPFS` or `WasmFS`? What about monkey-patching `WORKERFS`?

**Comment by tlively** [comment]
In WasmFS, the `readFile` JS API first reads the entire file into Wasm memory, then copies the contents out into JS. For very large files, it will not be possible to allocate that much Wasm memory, so `readFile` will fail.

You could indeed hack your own version of an OPFS file system by modifying `stream_ops` and `WORKERFS`, but I wouldn't recommend it.

I recommend using `read` instead of `readFile` to read the file contents in chunks rather than all at once.

**Comment by DenizUgur** [comment]
Thank you for your response. Indeed `readFile` wouldn't be the correct method here. Regardless the issue was related to seeking anyway. See my last edit please. This is a very edge case for me so I'll not pursue it anymore but just wanted to report it becasue the behaviour differs from WORKERFS and thought maybe this is something unexpected.

**Comment by tlively** [comment]
Hmm, seeking should use 64-bit values everywhere, so it should be able to handle very large files. What API were you calling where you saw problems with seeking?

**Comment by DenizUgur** [comment]
It's an educated guess. The application doesn't throw an exception and works properly with smaller offsets, so it's highly probably it's related to seeking or similar (not sure what else would be). I'm not using any JS API, yes I've used `-sFORCE_FILESYSTEM` but I think I don't have any need for it since browser APIs for `OPFS` is enough.

I'm calling POSIX APIs for file I/O, nothing emscripten related I guess.

**Comment by tlively** [comment]
Ok, sounds good. I'll close this issue, but feel free to reopen if it turns out that there is an Emscripten bug we can fix.

**Comment by DenizUgur** [comment]
Well, I'm not going to follow this issue because it's a very edge case. However, the problem with WasmFS is still present. If seeking with a value greater than `2^32-1` is something emscripten needs to handle then it needs to be checked. Maybe add a test case for this.

**Comment by sbc100** [comment]
I agree we should add test case for this if there is not one already.  @tlively do we have a test for seeking in large files beyond 2^32?   If not lets add one.

**Comment by brianhvo02** [comment]
I'd like to add my experiences with this issue. I've loaded an 8gb file into OPFS and mounted it with WasmFS. When reading over the file with ifstream it fails at 2^32. Has there been any progress on looking into this issue?

**Comment by brianhvo02** [comment]
I investigated a bit and it looks as if when reading a file larger than 4GB it only is able to read the first 2GB and the last 2GB. The offset remains the same though.

**Comment by brianhvo02** [comment]
@tlively Right now, the _wasmfs_opfs_read_blob and _wasmfs_opfs_get_size_blob functions in library_wasmfs_opfs.js are using uint32_t for file sizes and offset, which limits any operations to files less than 4GB. Using off_t would mean wrapping the offset and size in Number and BigInt constructors respectively. I'm thinking this could throw a wrench into people not using -sWASM_BIGINT so what's your take on this?

**Comment by tlively** [comment]
Sounds like those functions need to be updated to use i64s and bigints. IIUC, Emscripten has utilities to make that relatively painless, but I don't know exactly what the state of the art is there.

**Comment by goldwaving** [comment]
I'm running into this limit when working with large audio files (>2GB).

Looking at `opfs_backend.h`, it already uses `off_t` to set the size of the file (regardless of `WASM_BIGINT`), so changing the `read` and `write` functions to use `off_t` and `__i53abi` may work.  However I've noticed some odd behaviour with `__i53abi` that causes arguments following the `i53` argument to be invalid.  Reordering the arguments to put the `i53` argument last seems to work.

**Comment by DenizUgur** [comment]
Hi @sbc100, it's possible to reproduce this. See my latest comment above.

**Comment by sbc100** [comment]
> Hi @sbc100, it's possible to reproduce this. See my latest comment above.

Sorry, I was specifically referring to the `__i53abi` issue mentioned in https://github.com/emscripten-core/emscripten/issues/21335#issuecomment-2338966756.  Is that what you were referring to too?

**Comment by goldwaving** [comment]
I made a copy of the OPFS related files to create a test version to support >4GB files.  Here is a piece of the code for writing:
```c++
int _gwwasmfs_opfs_write_access(int access_id,
                              const uint8_t* buf,
                              uint32_t len,
                              off_t pos);
```
```js
  _gwwasmfs_opfs_write_access__i53abi: true,
  _gwwasmfs_opfs_write_access__deps: ['$gwwasmfsOPFSAccessHandles'],
  _gwwasmfs_opfs_write_access: {{{ asyncIf(!PTHREADS) }}} function(accessID, bufPtr, len, pos) {
    let accessHandle = gwwasmfsOPFSAccessHandles.get(accessID);
    let data = HEAPU8.subarray(bufPtr, bufPtr + len);
    console.log( "Write2: ", pos  );
```
This fails because when `pos` reaches 2^31, it goes negative, despite it being positive on the C++ side.  It seems like `__i53abi` had no effect.  I had to explicitly add a `__sig: 'iipij'` to avoid that.

For the reading code:
```c++
int _gwwasmfs_opfs_read_blob(em_proxying_ctx* ctx,
                           int blob_id,
                           uint8_t* buf,
                           uint32_t len,
                           off_t pos,
                           int32_t* nread);
```

```js
  _gwwasmfs_opfs_read_blob__i53abi: true,
  _gwwasmfs_opfs_read_blob__deps: ['$gwwasmfsOPFSBlobs', '$gwwasmfsOPFSProxyFinish'],
  _gwwasmfs_opfs_read_blob: async function(ctx, blobID, bufPtr, len, pos, nreadPtr) {
```
In this case, `nreadPtr` is always zero and you get a memory corruption error (and read error because the number of byte read cannot be passed back).  Again, it seems like `__i53abi` had no effect and it is using an `int32_t` for pos.  An explicit `__sig` was needed.

The `_wasmfs_opfs_set_size_file` function does not have an explicit `__sig`, so I don't know how (if?) that is working without it.

How are signatures determined when they are not given explicitly?  I've had strange problems where I had to specify it for some functions, but not for others.

The end result is that using `__i53abi` and explicit `__sig` for the read and write functions in OPFS allows it to support >4GB regardless of `WASM_BIGINT` (`_wasmfs_opfs_get_size_blob` also has to be changed to return the size like `wasmfs_opfs_get_size_file` does).

Maybe the documentation [here](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#calling-javascript-functions-as-function-pointers-from-c) could mention the `p` type and the `__i53abi` decorator with the `j` type and when an explict `__sig` is required.

**Comment by sbc100** [comment]
> The `_wasmfs_opfs_set_size_file` function does not have an explicit `__sig`, so I don't know how (if?) that is working without it.

We automatically generate `__sig` attributes for all builting JS functions.  See https://github.com/emscripten-core/emscripten/blob/3073806d3fde4320c81cb2dc7cf0e00378f52df1/src/library_sigs.js#L439 and https://github.com/emscripten-core/emscripten/blob/main/src/library_sigs.js#L1

I agree we should improve the documentation here.   I believe `__i53abi` doesn't do anything unless a signature with a `j` in it is found.    We could probably add a helper error message when its used without a `j` in the sig since that doesn't do anything.

**Comment by Elite** [comment]
Has anybody found a solution to get large files working as output to ffmpeg.wasm ?
